src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+use crate::edn::{Edn, Error, List, Map, Set, Vector};
+
+pub(crate) fn tokenize(edn: &str) -> std::str::Chars {
+    edn.chars()
+}
+
+pub(crate) fn parse(c: Option<char>, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    Ok(match c {
+        Some('[') => read_vec(chars)?,
+        Some('(') => read_list(chars)?,
+        Some('@') => read_set(chars)?,
+        Some('{') => read_map(chars)?,
+        edn => parse_edn(edn, chars)?,
+    })
+}
+
+pub(crate) fn parse_edn(c: Option<char>, chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    match c {
+        Some('\"') => Ok(read_str(chars)),
+        Some(':') => Ok(read_key(chars)),
+        Some(n) if n.is_numeric() => Ok(read_number(n, chars)?),
+        Some('-') => Ok(read_number('-', chars)?),
+        Some('\\') => Ok(read_char(chars)?),
+        Some(b) if b == 't' || b == 'f' || b == 'n' => Ok(read_bool_or_nil(b, chars)?),
+        a => Err(Error::ParseEdn(format!(
+            "{} could not be parsed",
+            a.unwrap().to_string()
+        ))),
+    }
+}
+
+fn read_key(chars: &mut std::str::Chars) -> Edn {
+    let c_len = chars
+        .clone()
+        .take_while(|c| !c.is_whitespace() && c != &')' && c != &']' && c != &'}')
+        .collect::<String>()
+        .len();
+    let mut key = String::from(":");
+    let key_chars = chars.take(c_len).collect::<String>();
+    key.push_str(&key_chars);
+    Edn::Key(key)
+}
+
+fn read_str(chars: &mut std::str::Chars) -> Edn {
+    let string = chars.take_while(|c| c != &'\"').collect::<String>();
```

Comment:
What happens if there's two backslashes (`\`) inside a String? :thinking:

Reply:
Actually thinking now you're looking for the `\"` character, so it probably isn't a problem what I am saying.

Reply:
Okay, I thought more, isn't this a problem when the character before this one is a `\`, because it's trying to skip the next slash? :thinking:

Reply:
if the char before is a `\` its char representation will be `\\` not `\`